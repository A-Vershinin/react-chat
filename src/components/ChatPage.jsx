import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';
import Sidebar from './Sidebar.jsx';
import ChatHeader from './ChatHeader.jsx';
import ChatMessages from './ChatMessages.jsx';
import ErrorMessage from './ErrorMessage.jsx';

const styles = theme => ({
  root: {
    position: 'relative',
    display: 'flex',
    width: '100%',
    height: '100%',
    backgroundColor: theme.palette.background.default,
  },
});

class ChatPage extends Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    match: PropTypes.shape({
      params: PropTypes.object.isRequired,
    }).isRequired,
    chats: PropTypes.shape({
      active: PropTypes.object,
      all: PropTypes.array.isRequired,
      my: PropTypes.array.isRequired,
    }).isRequired,
    activeUser: PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      username: PropTypes.string,
      isMember: PropTypes.bool.isRequired,
      isCreator: PropTypes.bool.isRequired,
      isChatMember: PropTypes.bool.isRequired,
    }).isRequired,
    messages: PropTypes.arrayOf(PropTypes.shape({
      chatId: PropTypes.string.isRequired,
      content: PropTypes.string.isRequired,
      sender: PropTypes.object.isRequired,
      createdAt: PropTypes.string.isRequired,
    })).isRequired,
    isConnected: PropTypes.bool.isRequired,
    sendMessage: PropTypes.func.isRequired,
    fetchAllChats: PropTypes.func.isRequired,
    fetchMyChats: PropTypes.func.isRequired,
    setActiveChat: PropTypes.func.isRequired,
    socketsConnect: PropTypes.func.isRequired,
    mountChat: PropTypes.func.isRequired,
    unmountChat: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
    createChat: PropTypes.func.isRequired,
    deleteChat: PropTypes.func.isRequired,
    joinChat: PropTypes.func.isRequired,
    leaveChat: PropTypes.func.isRequired,
    editUser: PropTypes.func.isRequired,
    error: PropTypes.instanceOf(Error),
  };

  static defaultProps = {
    error: null,
  };

  componentDidMount() {
    const {
      match, fetchAllChats, fetchMyChats, setActiveChat, socketsConnect, mountChat,
    } = this.props;

    Promise.all([fetchAllChats(), fetchMyChats()])
      .then(() => {
        socketsConnect();
      })
      .then(() => {
        const { chatId } = match.params;

        if (chatId) {
          setActiveChat(chatId);
          mountChat(chatId);
        }
      });
  }

  componentWillReceiveProps(nextProps) {
    const {
      match: { params },
      setActiveChat,
      unmountChat,
      mountChat,
    } = this.props;
    const { params: nextParams } = nextProps.match;

    // If we change route, then fetch messages from chat by chatID
    if (nextParams.chatId && params.chatId !== nextParams.chatId) {
      setActiveChat(nextParams.chatId);
      unmountChat(params.chatId);
      mountChat(nextParams.chatId);
    }
  }

  render() {
    const {
      classes,
      chats,
      logout,
      isConnected,
      createChat,
      deleteChat,
      joinChat,
      leaveChat,
      editUser,
      activeUser,
      messages,
      sendMessage,
      error,
    } = this.props;

    return (
      <div className={classes.root}>
        <ChatHeader
          activeUser={activeUser}
          logout={logout}
          isConnected={isConnected}
          activeChat={chats.active}
          deleteChat={deleteChat}
          leaveChat={leaveChat}
          editUser={editUser}
        />
        <Sidebar chats={chats} isConnected={isConnected} createChat={createChat} />
        <ChatMessages
          messages={messages}
          sendMessage={sendMessage}
          activeUser={activeUser}
          activeChat={chats.active}
          joinChat={joinChat}
          isConnected={isConnected}
        />
        <ErrorMessage error={error} />
      </div>
    );
  }
}

export default withRouter(withStyles(styles)(ChatPage));
