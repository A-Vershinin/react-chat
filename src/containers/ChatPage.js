import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import ChatPage from '../components/ChatPage.jsx';
import { fetchMyChats, fetchAllChats, fetchChat, setActiveChat, createChat, deleteChat, joinChat, leaveChat } from '../actions/chats';
import { logoutAction as logout } from '../actions/auth';
import { editUser } from '../actions/users';
import * as fromChats from '../reducers/chats';
import * as fromState from '../reducers';


function mapStateToProps(state) {
  const activeChat = fromChats.getById(state.chats, state.chats.activeId);
  // console.log("myChat", activeChat)
  return {
	   chats: {
       active: activeChat,
       all: fromChats.getByIds(state.chats, state.chats.allIds),
       my: fromChats.getByIds(state.chats, state.chats.myIds),
     },
     activeUser: {
        ...state.auth.user,
        isMember: fromState.isMember(state, activeChat),
        isCreator: fromState.isCreator(state, activeChat),
        isChatMember: fromState.isChatMember(state, activeChat)
     },
  }
};

const mapDispatchToProps = dispatch => bindActionCreators({
	fetchMyChats,
	fetchAllChats,
	fetchChat,
	setActiveChat,
  logout,
  createChat,
  deleteChat,
  joinChat,
  leaveChat,
  editUser,
}, dispatch);


export default connect(
	mapStateToProps,
	mapDispatchToProps
)(ChatPage)
