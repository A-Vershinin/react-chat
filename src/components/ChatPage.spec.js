/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter, Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import ChatPage from './ChatPage.jsx';

jest.mock('./Sidebar', () => () => 'Sidebar');
jest.mock('./ChatHeader', () => () => 'ChatHeader');
jest.mock('./ChatMessages', () => () => 'ChatMessages');
jest.mock('./ErrorMessage', () => () => 'ErrorMessage');

const mockProps = {
  classes: {
    root: 'some text',
  },
  isConnected: true,
  error: null,
  sendMessage: jest.fn(),
  fetchAllChats: jest.fn(),
  fetchMyChats: jest.fn(),
  setActiveChat: jest.fn(),
  socketsConnect: jest.fn(),
  mountChat: jest.fn(),
  unmountChat: jest.fn(),
  logout: jest.fn(),
  createChat: jest.fn(),
  deleteChat: jest.fn(),
  joinChat: jest.fn(),
  leaveChat: jest.fn(),
  editUser: jest.fn(),
  messages: [
    {
      chatId: '12345',
      content: 'some text',
      sender: {
        firstName: 'firstName',
        lastName: 'lastName',
        username: 'username',
        _id: '12345',
      },
      createdAt: '2018-03-16T10:53:23.200Z',
    },
  ],
  activeUser: {
    isMember: true,
    isCreator: true,
    isChatMember: true,
    firstName: 'firstName',
    lastName: 'lastName',
    username: 'username',
  },
  chats: {
    active: {},
    all: [],
    my: [],
  },
  match: {
    params: {
      chatId: '12345',
    },
  },
};

describe('<ChatPage />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter initialEntries={['/chat/123']}>
        <Route path="/chat/:chatId?" render={props => <ChatPage {...mockProps} {...props} />} />
      </MemoryRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/chat/123']}>
          <Route path="/chat/:chatId?" render={props => <ChatPage {...mockProps} {...props} />} />
              </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders when isConnected false', () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/chat/123']}>
          <Route path="/chat/:chatId?" render={props => <ChatPage {...mockProps} {...props} isConnected={false} />} />
              </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
