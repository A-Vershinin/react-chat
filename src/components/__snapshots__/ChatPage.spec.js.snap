// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatPage /> renders correctly 1`] = `
<div
  className="ChatPage-root-2 some text"
>
  ChatHeader
  Sidebar
  ChatMessages
  ErrorMessage
</div>
`;

exports[`<ChatPage /> renders when isConnected false 1`] = `
<div
  className="ChatPage-root-2 some text"
>
  ChatHeader
  Sidebar
  ChatMessages
  ErrorMessage
</div>
`;
